django/thunderstore/community/models/package_listing.py

```diff
@@ -342,6 +351,22 @@ def get_has_perms() -> bool:
                 )
             )
 
+        if not self.visibility:
+            raise ValidationError("Insufficient permissions to view")
+
+        if user is not None:
+            if not self.visibility.public_detail:
+                if not (
+                    self.visibility.owner_detail
+                    and self.package.owner.can_user_access(user)
+                ):
+                    if not (
+                        self.visibility.moderator_detail
+                        and self.community.can_user_manage_packages(user)
+                    ):
+                        if not (self.visibility.admin_detail and user.is_superuser):
+                            raise ValidationError("Insufficient permissions to view")
```

Comment:
Rather than this long piece of logic, ideally we should add a function to the `VisibilityMixin` that behaves akin to the `VisibilityQuerySet.visible_detail` function.

Granted, in reality I think the `ensure_can_be_viewed_by_user` function calls themselves should be replaced by something in the `VisibilityMixin`, but there's no need to go that far in this PR yet

Reply:
Addressed in [Refactor can_be_viewed_by_user](https://github.com/thunderstore-io/Thunderstore/pull/1073/commits/a94d46415c6328351923b3fde8b978ace31d054a)

It no longer uses the validation paradigm, but it can be added back if it's important.